### ISSUE TITLE
Thread stuck on HA_SPIN_LOCK in dns_session_release()

### ISSUE THREAD

**Issue by Feder1co5oave** [haproxy/haproxy#2277]
### Detailed Description of the Problem

HAProxy exits after seconds/minutes with error:
```
*>Thread 1 : id=0x7fc3cc757ec0 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=1 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=126077523 now=3009464544 diff=2883387021
             curr_task=0x561ef0a900f0 (task) calls=387 last=0
               fct=0x561eef9d6eca(task_run_applet) ctx=0x561ef0a8ffe0(<STRMDNS>)
             strm=0x561ef0a90290,46 src=<STRMDNS> fe=kube-dns be=kube-dns dst=coredns
             txn=(nil),0 txn.req=-,0 txn.rsp=-,0
             rqf=848020 rqa=0 rpf=8004a020 rpa=0 sif=EST,284048 sib=EST,2800d8
             af=0x561ef0a8ffe0,0 csf=(nil),0
             ab=(nil),0 csb=0x561ef0a90b10,9000
             cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0)
             cob=0x561ef0a90920,10042300:PASS(0x561ef0a90b60)/RAW((nil))/tcpv4(22)
             call trace(23):
             | 0x561eefa54c6d [eb 11 90 eb 01 90 8b 45]: wdt_handler+0x16f/0x184
             | 0x7fc3ccee2140 [48 c7 c0 0f 00 00 00 0f]: libpthread:+0x13140
             | 0x561eef8a2ec7 [48 83 e8 80 48 89 45 d8]: main+0x386cf
             | 0x561eef8804fe [90 c9 c3 55 48 89 e5 48]: main+0x15d06
             | 0x561eef88372e [48 8b 45 e8 c6 00 09 48]: main+0x18f36
             | 0x561eef8a0721 [90 c9 c3 55 48 89 e5 48]: main+0x35f29
             | 0x561eef8a2c4a [48 8b 45 a8 48 89 c7 e8]: main+0x38452
             | 0x561eef9d6f86 [48 8b 45 e0 48 89 c7 e8]: task_run_applet+0xbc/0x284
  Thread 2 : id=0x7fc3cc74c700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=110956955 now=113552752 diff=2595797
             curr_task=0
  Thread 3 : id=0x7fc3cabb9700 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=111062551 now=1996681788 diff=1885619237
             curr_task=0x561ef086e1a0 (task) calls=129 last=0
               fct=0x561eef8a3b9b(main+0x393a3) ctx=0x561ef0909af0
  Thread 4 : id=0x7fc3ca3b8700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=109876257 now=112500478 diff=2624221
             curr_task=0
  Thread 5 : id=0x7fc3c9bb7700 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
             stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=111316365 now=174774979 diff=63458614
             curr_task=0x561ef090ac60 (task) calls=26 last=0
               fct=0x561eef8a365a(main+0x38e62) ctx=0x561ef090aab0
```
**thread 1 is stuck on something regarding a DNS (TCP) session.**

### Steps to Reproduce the Behavior

**I encountered this on haproxy 2.4.24, was able to reproduce on 2.4.23, although 2.4.22 does NOT seem to be affected.**

I'm running HAProxy in containerd (Kubernetes), with a DNS resolver with address `tcp@10.106.0.10:53`.

The instance is receiving virtually no requests, apart from liveness probes.

For some reason this seems to happen with coredns v1.9.3, but not with v1.10.1.

------------------------

I recompiled sources from [2.4.23](http://www.haproxy.org/download/2.4/src/haproxy-2.4.23.tar.gz) with debug symbols (`DEBUG="-DDEBUG_MEMORY_POOLS -DDEBUG_STRICT=2 -DDEBUG_STRICT_ACTION=3" DEBUG_CFLAGS="-ggdb3" CPU_CFLAGS="-O0"`),
reproduced the error and inspected the corefile with gdb. Here's some output:

```
gdb /usr/local/sbin/haproxy /tmp/core.35
(gdb) dir /tmp/haproxy-2.4.23
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fc3cc8c2537 in __GI_abort () at abort.c:79
#2  0x0000561eef8bf5b7 in ha_panic () at src/debug.c:334
#3  0x0000561eefa54c6d in wdt_handler (sig=14, si=0x7ffcde4a32b0, arg=0x7ffcde4a3180) at src/wdt.c:122
#4  <signal handler called>
#5  0x0000561eef8a2ec7 in dns_session_release (appctx=0x561ef0a8ffe0) at src/dns.c:821
#6  0x0000561eef8804fe in si_applet_release (si=0x561ef0a90540) at include/haproxy/stream_interface.h:236
#7  0x0000561eef88372e in stream_int_shutr_applet (si=0x561ef0a90540) at src/stream_interface.c:1699
#8  0x0000561eef8a0721 in si_shutr (si=0x561ef0a90540) at include/haproxy/stream_interface.h:431
#9  0x0000561eef8a2c4a in dns_session_io_handler (appctx=0x561ef0a8ffe0) at src/dns.c:765
#10 0x0000561eef9d6f86 in task_run_applet (t=0x561ef0a900f0, context=0x561ef0a8ffe0, state=16388) at src/applet.c:90
#11 0x0000561eef913f7c in run_tasks_from_lists (budgets=0x7ffcde4a4130) at src/task.c:599
#12 0x0000561eef91498f in process_runnable_tasks () at src/task.c:853
#13 0x0000561eef86a112 in run_poll_loop () at src/haproxy.c:2655
#14 0x0000561eef86a671 in run_thread_poll_loop (data=0x0) at src/haproxy.c:2830
#15 0x0000561eef86c132 in main (argc=5, argv=0x7ffcde4a4558) at src/haproxy.c:3552
```

Error seems to have been originated in dns_session_release(), when trying to acquire lock
https://github.com/haproxy/haproxy/blob/62cb999fe65e415ba7176354172612a7ee8bdf04/src/dns.c#L821
```
(gdb) frame 5 
#5  0x0000561eef8a2ec7 in dns_session_release (appctx=0x561ef0a8ffe0) at src/dns.c:821
821		HA_SPIN_LOCK(DNS_LOCK, &dss->lock);
(gdb) p *dss
$1 = {srv = 0x561ef0909b50, ring_req = 0x561ef090ab40, max_slots = 0, maxconn = 0, idle_conns = 1, cur_conns = 1, max_active_conns = 1, ofs_req = 10496, ofs_rsp = 0, task_req = 0x561ef086e1a0, 
  task_rsp = 0x561ef090ab90, task_idle = 0x561ef090ac60, free_sess = {n = 0x561ef090ab00, p = 0x561ef090ab00}, idle_sess = {n = 0x561ef0a6fde0, p = 0x561ef0a6fde0}, wait_sess = {
    n = 0x561ef090ab20, p = 0x561ef090ab20}, lock = 4294967300}
(gdb) p *ds
$2 = {appctx = 0x561ef0a8ffe0, dss = 0x561ef090aab0, tx_msg_offset = 0, nb_queries = 0, onfly_queries = 0, query_counter = 128, list = {n = 0x561ef090ab10, p = 0x561ef090ab10}, waiter = {
    n = 0x561ef0a6fdf0, p = 0x561ef0a6fdf0}, queries = {n = 0x561ef0a6fe00, p = 0x561ef0a6fe00}, task_exp = 0x561ef0a8ff10, query_ids = {b = {0x0, 0x1}}, ofs = 10496, ring = {buf = {
      size = 65540, area = 0x561ef0a7fee0 "", data = 10497, head = 0}, ofs = 0, waiters = {n = 0x561ef0a90070, p = 0x561ef0a90070}, lock = 0, readers_count = 1}, rx_msg = {len = 0, offset = 0, 
    area = 0x561ef0a6feb0 "\023\f\205"}, tx_ring_area = 0x561ef0a7fee0 "", shutdown = 0}
```

The DNS session shutdown originated in dns_session_io_handler, after "close:" jump
https://github.com/haproxy/haproxy/blob/62cb999fe65e415ba7176354172612a7ee8bdf04/src/dns.c#L763-L766
```
(gdb) fram 9
#9  0x0000561eef8a2c4a in dns_session_io_handler (appctx=0x561ef0a8ffe0) at src/dns.c:765
(gdb) p stopping
$5 = 0
(gdb) p ds->shutdown 
$6 = 0
(gdb) p *si
$7 = {state = SI_ST_EST, prev_state = SI_ST_EST, flags = 2637896, end = 0x561ef0a8ffe0, ops = 0x561eefb80840 <si_applet_ops>, exp = 0, err_type = 0, conn_retries = 0, hcto = 0, 
  wait_event = {tasklet = 0x561ef0a907d0, events = 0}, l7_buffer = {size = 0, area = 0x0, data = 0, head = 0}}
(gdb) p ds->rx_msg.len
$8 = 0
(gdb) p ret
$9 = -1
```
The jump to close might have originated from any one of lines 663, 684, 706.
The jump at line 755 is preceded by an unlock() so my guess is it could not have caused this.

The causing commit is probably https://github.com/haproxy/haproxy/commit/cf537f68771ae4105e41800942fec4b2693a7a82
which was introduced between 2.4.22 and 2.4.23.
/cc @capflam 

If that lock is not reentrant (I think it is not), it is already locked in
https://github.com/haproxy/haproxy/blob/62cb999fe65e415ba7176354172612a7ee8bdf04/src/dns.c#L650
and then in
https://github.com/haproxy/haproxy/blob/62cb999fe65e415ba7176354172612a7ee8bdf04/src/dns.c#L821
a possible fix could be to unlock it before closing the DNS session.

### Configuration

```haproxy
defaults
  timeout client 1s
  timeout client-fin 5s
  timeout http-request 1s
  timeout queue 200
  timeout server 500
  timeout server-fin 5s
  timeout connect 1s
  timeout http-keep-alive 15s
  backlog 10000

  balance leastconn
  option httpchk GET /status
  mode http
  option http-keep-alive
  option http-use-htx
  #option abortonclose
  option redispatch
  http-reuse always
  max-keep-alive-queue 50
  option httplog
  option log-separate-errors

global
  maxconn 60000
  daemon
  nbproc 1
  nbthread 5
  cpu-map auto:1/2-5 0-3
  tune.listener.multi-queue on

frontend http
  bind :8000 tfo process 1/2-
  option tcp-smart-accept
  maxconn 50000
  default_backend pbs

backend pbs
  retries 1
  default-server weight 100 resolvers kube-dns resolve-prefer ipv4 resolve-opts prevent-dup-ip,ignore-weight check inter 500 downinter 1s maxconn 300 maxqueue 10 pool-low-conn 20
  server-template pbs    60 _http._tcp.pbs.default.svc.cluster.local slowstart 5s

resolvers kube-dns
  nameserver coredns tcp@10.106.0.10:53
  resolve_retries 5
  accepted_payload_size 65535
  hold valid 5s
  hold obsolete 11s
```

### Output of `haproxy -vv`

```plain
## haproxy -vv

HAProxy version 2.4.23-62cb999 2023/06/09 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.23.html
Running on: Linux 5.15.0-75-generic #82-Ubuntu SMP Tue Jun 6 23:10:23 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O0 -ggdb3 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_MEMORY_POOLS -DDEBUG_STRICT=2 -DDEBUG_STRICT_ACTION=3

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL +EPOLL -EVPORTS +FUTEX +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PRIVATE_CACHE -PROCCTL +PROMEX -PTHREAD_PSHARED -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=24).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

## uname -a
`Linux 13417e50302f 5.15.0-75-generic #82-Ubuntu SMP Tue Jun 6 23:10:23 UTC 2023 x86_64 GNU/Linux`

Processor is usually Ryzen 9 5900x with `performance` governor.

### Last Outputs and Backtraces

<details>
<summary>Output of t a a bt full</summary>
<pre><code>
Thread 5 (Thread 0x7fc3ca3b8700 (LWP 38)):
warning: Section `.reg-xstate/38' in core file too small.
#0  0x00007fc3cc99bd56 in epoll_wait (epfd=16, events=0x7fc3c0025660, maxevents=200, timeout=14) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
No locals.
#1  0x0000561eef60a959 in _do_poll (p=0x561eefb895c0 <cur_poller>, exp=1576347630, wake=0) at src/ev_epoll.c:195
        timeout = 14
        status = 22046
        fd = -1
        count = -258458016
        updt_idx = 0
        wait_time = 14
        old_fd = -1
#2  0x0000561eef86a3af in run_poll_loop () at src/haproxy.c:2712
        next = 1576347630
        wake = 0
#3  0x0000561eef86a671 in run_thread_poll_loop (data=0x3) at src/haproxy.c:2830
        ptaf = 0x561eefb88430 <per_thread_alloc_list>
        ptif = 0x561eefb88440 <per_thread_init_list>
        ptdf = 0x0
        ptff = 0x0
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 20, __wseq32 = {__low = 20, __high = 0}}, {__g1_start = 12, __g1_start32 = {__low = 12, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 16, __wrefs = 0, __g_signals = {0, 0}}, __size = "\024\000\000\000\000\000\000\000\f", '\000' <repeats 23 times>, "\020", '\000' <repeats 14 times>, __align = 20}
#4  0x00007fc3cced6ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140478888249088, 7913066827411472024, 140724037894606, 140724037894607, 140478888204352, 8396800, -7901767654014599528, -7901770603272356200}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x00007fc3cc99ba2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 4 (Thread 0x7fc3c9bb7700 (LWP 39)):
warning: Section `.reg-xstate/39' in core file too small.
#0  0x0000561eef8a3688 in dns_process_idle_exp (t=0x561ef090ac60, context=0x561ef090aab0, state=524) at src/dns.c:996
        __pl_l = 0x561ef090ab30
        __pl_r = 4294967300
        dss = 0x561ef090aab0
        ds = 0x9
        dsb = 0x561ef0983b20
        target = 0
        cur_active_conns = 0
#1  0x0000561eef913f7c in run_tasks_from_lists (budgets=0x7fc3c9bac660) at src/task.c:599
        process = 0x561eef8a365a <dns_process_idle_exp>
        tl_queues = 0x561eefc394c0 <task_per_thread+1088>
        t = 0x561ef090ac60
        budget_mask = 15 '\017'
        profile_entry = 0x0
        done = 0
        queue = 1
--Type <RET> for more, q to quit, c to continue without paging--c
        state = 524
        ctx = 0x561ef090aab0
#2  0x0000561eef91498f in process_runnable_tasks () at src/task.c:853
        tt = 0x561eefc39480 <task_per_thread+1024>
        lrq = 0x0
        grq = 0x0
        t = 0x561ef090ac60
        default_weights = {64, 48, 16, 1}
        max = {0, 139, 0, 0}
        max_total = 48
        tmp_list = 0x0
        queue = 4
        max_processed = 140
        lpicked = 0
        gpicked = 1
        heavy_queued = 1
        budget = 140
#3  0x0000561eef86a112 in run_poll_loop () at src/haproxy.c:2655
        next = 1576347553
        wake = 0
#4  0x0000561eef86a671 in run_thread_poll_loop (data=0x4) at src/haproxy.c:2830
        ptaf = 0x561eefb88430 <per_thread_alloc_list>
        ptif = 0x561eefb88440 <per_thread_init_list>
        ptdf = 0x0
        ptff = 0x0
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 20, __wseq32 = {__low = 20, __high = 0}}, {__g1_start = 12, __g1_start32 = {__low = 12, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 16, __wrefs = 0, __g_signals = {0, 0}}, __size = "\024\000\000\000\000\000\000\000\f", '\000' <repeats 23 times>, "\020", '\000' <repeats 14 times>, __align = 20}
#5  0x00007fc3cced6ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140478879856384, 7913066827411472024, 140724037894606, 140724037894607, 140478879811648, 8396800, -7901759956896334184, -7901770603272356200}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#6  0x00007fc3cc99ba2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 3 (Thread 0x7fc3cc74c700 (LWP 36)):
warning: Section `.reg-xstate/36' in core file too small.
#0  0x00007fc3cc99bd56 in epoll_wait (epfd=10, events=0x7fc3c4025d50, maxevents=200, timeout=14) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
No locals.
#1  0x0000561eef60a959 in _do_poll (p=0x561eefb895c0 <cur_poller>, exp=1576347630, wake=0) at src/ev_epoll.c:195
        timeout = 14
        status = 22046
        fd = -1
        count = -258458016
        updt_idx = 0
        wait_time = 14
        old_fd = -1
#2  0x0000561eef86a3af in run_poll_loop () at src/haproxy.c:2712
        next = 1576347630
        wake = 0
#3  0x0000561eef86a671 in run_thread_poll_loop (data=0x1) at src/haproxy.c:2830
        ptaf = 0x561eefb88430 <per_thread_alloc_list>
        ptif = 0x561eefb88440 <per_thread_init_list>
        ptdf = 0x7fc3cc741840
        ptff = 0x7ffcde4a436f
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 20, __wseq32 = {__low = 20, __high = 0}}, {__g1_start = 12, __g1_start32 = {__low = 12, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 16, __wrefs = 0, __g_signals = {0, 0}}, __size = "\024\000\000\000\000\000\000\000\f", '\000' <repeats 23 times>, "\020", '\000' <repeats 14 times>, __align = 20}
#4  0x00007fc3cced6ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140478925555456, 7913066827411472024, 140724037894606, 140724037894607, 140478925510720, 94690884309392, -7901771487272911208, -7901770603272356200}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x00007fc3cc99ba2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 2 (Thread 0x7fc3cabb9700 (LWP 37)):
warning: Section `.reg-xstate/37' in core file too small.
#0  0x0000561eef8a3bfd in dns_process_req (t=0x561ef086e1a0, context=0x561ef0909af0, state=4108) at src/dns.c:1108
        __pl_r = 4294967300
        ns = 0x561ef0909af0
        dss = 0x561ef090aab0
        ring = 0x561ef090ab40
        buf = 0x561ef090ab40
        msg_len = 0
        len = 94690885124720
        cnt = 2
        ofs = 94690885124720
        ds = 0x561ef0924270
        ads = 0x278
#1  0x0000561eef913f7c in run_tasks_from_lists (budgets=0x7fc3cabae660) at src/task.c:599
        process = 0x561eef8a3b9b <dns_process_req>
        tl_queues = 0x561eefc392c0 <task_per_thread+576>
        t = 0x561ef086e1a0
        budget_mask = 15 '\017'
        profile_entry = 0x0
        done = 0
        queue = 1
        state = 4108
        ctx = 0x561ef0909af0
#2  0x0000561eef91498f in process_runnable_tasks () at src/task.c:853
        tt = 0x561eefc39280 <task_per_thread+512>
        lrq = 0x0
        grq = 0x0
        t = 0x561ef086e1a0
        default_weights = {64, 48, 16, 1}
        max = {0, 136, 0, 0}
        max_total = 48
        tmp_list = 0x0
        queue = 4
        max_processed = 137
        lpicked = 0
        gpicked = 1
        heavy_queued = 1
        budget = 137
#3  0x0000561eef86a112 in run_poll_loop () at src/haproxy.c:2655
        next = 1576345730
        wake = 0
#4  0x0000561eef86a671 in run_thread_poll_loop (data=0x2) at src/haproxy.c:2830
        ptaf = 0x561eefb88430 <per_thread_alloc_list>
        ptif = 0x561eefb88440 <per_thread_init_list>
        ptdf = 0x0
        ptff = 0x0
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 20, __wseq32 = {__low = 20, __high = 0}}, {__g1_start = 12, __g1_start32 = {__low = 12, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 16, __wrefs = 0, __g_signals = {0, 0}}, __size = "\024\000\000\000\000\000\000\000\f", '\000' <repeats 23 times>, "\020", '\000' <repeats 14 times>, __align = 20}
#5  0x00007fc3cced6ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140478896641792, 7913066827411472024, 140724037894606, 140724037894607, 140478896597056, 8396800, -7901766552892359016, -7901770603272356200}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#6  0x00007fc3cc99ba2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 1 (Thread 0x7fc3cc757ec0 (LWP 35)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {8192, 0 <repeats 15 times>}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007fc3cc8c2537 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0xffffffffffffffff, sa_sigaction = 0xffffffffffffffff}, sa_mask = {__val = {18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 94690871611720, 4981100279980124672, 266, 94690869263586, 0, 98784247808, 4294967014, 140478933504004, 140724037890304}}, sa_flags = 2491, sa_restorer = 0x561eefc41148 <ha_thread_info+8>}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x0000561eef8bf5b7 in ha_panic () at src/debug.c:334
No locals.
#3  0x0000561eefa54c6d in wdt_handler (sig=14, si=0x7ffcde4a32b0, arg=0x7ffcde4a3180) at src/wdt.c:122
        n = 3009428787
        p = 126077523
        thr = 0
#4  <signal handler called>
No locals.
#5  0x0000561eef8a2ec7 in dns_session_release (appctx=0x561ef0a8ffe0) at src/dns.c:821
        __pl_l = 0x561ef090ab30
        __pl_r = 4294967300
        ds = 0x561ef0a6fdc0
        dss = 0x561ef090aab0
        __pl_r = <optimized out>
        __pl_i = <optimized out>
        ret = <optimized out>
        __pl_i = <optimized out>
#6  0x0000561eef8804fe in si_applet_release (si=0x561ef0a90540) at include/haproxy/stream_interface.h:236
        appctx = 0x561ef0a8ffe0
#7  0x0000561eef88372e in stream_int_shutr_applet (si=0x561ef0a90540) at src/stream_interface.c:1699
        ic = 0x561ef0a902a0
#8  0x0000561eef8a0721 in si_shutr (si=0x561ef0a90540) at include/haproxy/stream_interface.h:431
No locals.
#9  0x0000561eef8a2c4a in dns_session_io_handler (appctx=0x561ef0a8ffe0) at src/dns.c:765
        si = 0x561ef0a90540
        ds = 0x561ef0a6fdc0
        ring = 0x561ef0a6fe30
        buf = 0x561ef0a6fe30
        msg_len = 0
        available_room = 32764
        len = 94690869078790
        cnt = 8024375168
        ofs = 94690886615088
        ret = -1
        __pl_r = <optimized out>
        __pl_i = <optimized out>
        ret = <optimized out>
        __pl_i = <optimized out>
        __pl_r = <optimized out>
        __pl_i = <optimized out>
        ret = <optimized out>
        __pl_i = <optimized out>
#10 0x0000561eef9d6f86 in task_run_applet (t=0x561ef0a900f0, context=0x561ef0a8ffe0, state=16388) at src/applet.c:90
        app = 0x561ef0a8ffe0
        si = 0x561ef0a90540
        rate = 3
        count = 0
#11 0x0000561eef913f7c in run_tasks_from_lists (budgets=0x7ffcde4a4130) at src/task.c:599
        process = 0x561eef9d6eca <task_run_applet>
        tl_queues = 0x561eefc390c0 <task_per_thread+64>
        t = 0x561ef0a900f0
        budget_mask = 15 '\017'
        profile_entry = 0x0
        done = 0
        queue = 1
        state = 16388
        ctx = 0x561ef0a8ffe0
#12 0x0000561eef91498f in process_runnable_tasks () at src/task.c:853
        tt = 0x561eefc39080 <task_per_thread>
        lrq = 0x0
        grq = 0x0
        t = 0x561ef0a900f0
        default_weights = {64, 48, 16, 1}
        max = {0, 134, 0, 0}
        max_total = 48
        tmp_list = 0x0
        queue = 4
        max_processed = 135
        lpicked = 1
        gpicked = 0
        heavy_queued = 1
        budget = 135
#13 0x0000561eef86a112 in run_poll_loop () at src/haproxy.c:2655
        next = 1576344747
        wake = 0
#14 0x0000561eef86a671 in run_thread_poll_loop (data=0x0) at src/haproxy.c:2830
        ptaf = 0x561eefb88430 <per_thread_alloc_list>
        ptif = 0x561eefb88440 <per_thread_init_list>
        ptdf = 0x0
        ptff = 0x0
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 20, __wseq32 = {__low = 20, __high = 0}}, {__g1_start = 12, __g1_start32 = {__low = 12, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 16, __wrefs = 0, __g_signals = {0, 0}}, __size = "\024\000\000\000\000\000\000\000\f", '\000' <repeats 23 times>, "\020", '\000' <repeats 14 times>, __align = 20}
#15 0x0000561eef86c132 in main (argc=5, argv=0x7ffcde4a4558) at src/haproxy.c:3552
        blocked_sig = {__val = {18446744067199990583, 18446744073709551615 <repeats 15 times>}}
        old_sig = {__val = {0, 140478927490572, 140724037895184, 0, 1, 279172874304, 48, 32, 94686849007619, 140724037895192, 206158430224, 0, 206158430211, 532575944814, 0, 511101108315}}
        i = 5
        err = 0
        retry = 200
        limit = {rlim_cur = 120098, rlim_max = 2000000}
        pidfd = -1
        intovf = 6
</code></pre>
</details>

**Comment by capflam** [comment]
Indeed, It seems you're right. There is no lock on upper versions but DSS must be unlock on the 2.4. Tanks ! I will fix it soon.

**Comment by capflam** [comment]
I pushed a fix to 2.4. It should be ok now. Many thanks for the analysis !

**Comment by Feder1co5oave** [comment]
@capflam I have recompiled with the fix from
https://git.haproxy.org/?p=haproxy-2.4.git;a=snapshot;h=70ab149fe9c64b8173688042944932aa6615b1e0;sf=tgz
and haven't been able to reproduce the crash. This should be ok then.

Thanks so much!
